src/plugins/block.ts

```diff
@@ -0,0 +1,216 @@
+import {
+  Assignment,
+  Block,
+  block,
+  Expr,
+  Identifier,
+  manyToManyAssignment,
+  Node,
+  oneToManyAssignment,
+  VarDeclaration,
+  varDeclarationBlock,
+  VarDeclarationWithAssignment,
+  varDeclarationWithAssignment,
+} from "../IR";
+import { Plugin } from "../common/Language";
+import { Spine } from "../common/Spine";
+import { stringify } from "../common/stringify";
+
+/**
+ * Collects neighbouring block children matching a predicate and replaces them with a different set of children.
+ * @param collectPredicate Condition for expr to be collected.
+ * @param transform Transforming function.
+ * @param blockPredicate Condition for block to consider its children.
+ */
+export function blockChildrenCollectAndReplace<T extends Expr = Expr>(
+  name: string,
+  collectPredicate: (expr: Expr, spine: Spine<Expr>, previous: T[]) => boolean,
+  transform: (exprs: T[]) => Expr[],
+  blockPredicate: (block: Block, spine: Spine<Block>) => boolean = () => true
+): Plugin {
+  return {
+    name,
+    visit(node, spine) {
+      if (
+        node.kind === "Block" &&
+        blockPredicate(node, spine as Spine<Block>)
+      ) {
+        const newNodes: Expr[] = [];
+        let changed = false;
+        let collected: T[] = [];
+        for (const childSpine of spine.getChildSpines()) {
+          const expr = childSpine.node as Expr;
+          if (collectPredicate(expr, childSpine as Spine<Expr>, collected)) {
+            collected.push(expr as any as T);
+          } else if (collectPredicate(expr, childSpine as Spine<Expr>, [])) {
+            if (collected.length > 1) {
+              newNodes.push(...transform(collected));
+              changed = true;
+            } else {
+              newNodes.push(...collected);
+            }
+            collected = [expr as any as T];
```

Comment:
`as any` is unnecessary because T is a narrower type to Expr.

```suggestion
            collected.push(expr as T);
          } else if (collectPredicate(expr, childSpine as Spine<Expr>, [])) {
            if (collected.length > 1) {
              newNodes.push(...transform(collected));
              changed = true;
            } else {
              newNodes.push(...collected);
            }
            collected = [expr as T];
```